### PR TITLE
feat: show warning notification on `drive.mount`

### PR DESCRIPTION
**Pull request by hjjackyang**
* **Why**? While we work on implementing `drive.mount` for VS Code extension, this is a stop-gap solution to point users to our [workaround doc](https://github.com/googlecolab/colab-vscode/wiki/Known-Issues-and-Workarounds#drivemount), and also to let them know there is already an [open issue](https://github.com/googlecolab/colab-vscode/issues/256), so they don't file another one.
* [Screencast](https://screencast.googleplex.com/cast/NTI2MDg3MjA0NDc3MzM3Nnw5ZTg5MmFiOS0wNg)
* [Bug reference](https://b.corp.google.com/issues/461876404)
* Notes:
   * I initially experimented with faking an `execute_reply` message and displaying the warning as a kernel execution error. However, that didn't work out as how I expected, hence the current solution to show as a VS Code warning notification.
   * For code organization, I put this new warning logic in a separate file, so it's easier to 1) unit test, and 2) deprecate after we implement `drive.mount` for VS Code.

Preview of the warning notification message:
Drive mount warning

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -195,10 +195,12 @@
   "dependencies": {
     "glob": "^11.1.0",
     "google-auth-library": "^9.15.0",
+    "isomorphic-ws": "^4.0.1",
```
Comment (kevineger) [inline package.json]: Do we actually need this?

As I understand this package exists for isomorphism between `ws` in Node and `WebSocket` in browsers.

Our extension is a standard extension, which always runs on the VS Code server - https://code.visualstudio.com/docs/remote/vscode-server#_architecture. We don't publish a "Web Extension" - https://code.visualstudio.com/api/extension-guides/web-extensions.

Since we're never running in the browser, I don't _think_ we need compatibility with `WebSocket`. In other words, I think we can purely use `ws`. Sorry, this may require a bit of homework / research.

File: src/jupyter/assignments.ts [hunk 2]
```diff
@@ -217,6 +219,10 @@ export class AssignmentManager implements vscode.Disposable {
         connectionInformation: {
           ...server.connectionInformation,
           fetch: colabProxyFetch(server.connectionInformation.token),
+          WebSocket: colabProxyWebSocket(
+            this.vs,
+            server.connectionInformation.token,
+          ),
         },
       }));
```
Comment (kevineger) [inline src/jupyter/assignments.ts]: Totally optional nit, this may read easier.

```ts
.map((server) => {
  const c = server.connectionInformation;
  return {
    ...server,
    connectionInformation: {
      ...c,
      fetch: colabProxyFetch(c.token),
      WebSocket: colabProxyWebSocket(this.vs, c.token),
    },
  };
});
```

File: src/jupyter/assignments.ts [hunk 3]
```diff
@@ -651,3 +658,74 @@ function colabProxyFetch(
...
+    };
+  };
+
+  return class OurWebSocket extends WebSocketIsomorphic {
+    constructor(
+      url: string | URL,
+      protocols?: string | string[],
+      options?: WebSocketIsomorphic.ClientOptions,
+    ) {
+      super(url, protocols, addOurHeaders(options));
```
Comment (kevineger) [inline src/jupyter/assignments.ts]: _Our_ -> _Colab_

```suggestion
  // These custom headers are required for Colab's proxy WebSocket to work.
  const colabHeaders: Record<string, string> = {};
  colabHeaders[COLAB_RUNTIME_PROXY_TOKEN_HEADER.key] = token;
  colabHeaders[COLAB_CLIENT_AGENT_HEADER.key] = COLAB_CLIENT_AGENT_HEADER.value;

  const addColabHeaders = (
    options?: WebSocketIsomorphic.ClientOptions,
  ): WebSocketIsomorphic.ClientOptions => {
    options ??= {};
    options.headers ??= {};
    const headers: Record<string, string> = {
      ...options.headers,
      ...colabHeaders,
    };
    return {
      ...options,
      headers,
    };
  };

  return class ColabWebSocket extends WebSocketIsomorphic {
    constructor(
      url: string | URL,
      protocols?: string | string[],
      options?: WebSocketIsomorphic.ClientOptions,
    ) {
      super(url, protocols, addColabHeaders(options));
```

File: src/jupyter/assignments.ts [hunk 4]
```diff
@@ -651,3 +658,74 @@ function colabProxyFetch(
...
+    options ??= {};
+    options.headers ??= {};
+    const headers: Record<string, string> = {
+      ...options.headers,
+      ...colabHeaders,
+    };
+    return {
+      ...options,
+      headers,
+    };
```
Comment (kevineger) [inline src/jupyter/assignments.ts]: Totally optional nit, single line for readability.

```suggestion
    return { ...options, headers };
```

File: src/jupyter/assignments.ts [hunk 5]
```diff
@@ -651,3 +658,74 @@ function colabProxyFetch(
...
+    override send(
+      data: string,
+      options: {
+        mask?: boolean;
+        binary?: boolean;
+        compress?: boolean;
+        fin?: boolean;
+      },
+      cb?: (err?: Error) => void,
+    ): void;
```
Comment (kevineger) [inline src/jupyter/assignments.ts]: I'm pretty sure we don't need all of these overloads re-defined. Overriding only the implementation signature is sufficient.

The only reason I think we'd want to override the overloads is if we handled `OurWebSocket` directly in code and wanted the simplified call signatures. But considering we don't, just the impl. sig should be enough.

File: src/jupyter/assignments.ts [hunk 6]
```diff
@@ -651,3 +658,74 @@ function colabProxyFetch(
...
+            compress?: boolean;
+            fin?: boolean;
+          }
+        | ((err?: Error) => void)
+        | undefined,
+      cb?: (err?: Error) => void,
+    ) {
+      void warnOnDriveMount(vs, data);
+
+      // This is required to override all `send` overload signatures:
```
Comment (kevineger) [inline src/jupyter/assignments.ts]: Nit: given the signature, if you didn't have this it'd complain no? Optionally consider removing comment.

_Argument of type [...] is not assignable to parameter of type [...]_

File: src/jupyter/drive-mount-warning.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  };
+  if (
+    parsedJupyterMessage.header.msg_type === 'execute_request' &&
+    DRIVE_MOUNT_PATTERN.exec(parsedJupyterMessage.content.code)
+  ) {
+    await notifyDriveMountUnsupported(vs);
+  }
+}
+
+async function notifyDriveMountUnsupported(vs: typeof vscode): Promise<void> {
```
Comment (kevineger) [inline src/jupyter/drive-mount-warning.ts]: Considering this function renders the warning message and opens external links, we'd never care to wait for the result. As such, it probably makes more sense to have this be a sync method to simplify callsites. As is, the `void` call requires readers to click in a figure out what async work is being done, to ensure sync mechanism aren't needed.

In other words, consider making this a synchronous function and chaining the selected action logic as a thenable.

File: src/jupyter/drive-mount-warning.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    case DriveMountUnsupportedAction.VIEW_WORKAROUND:
+      vs.env.openExternal(vs.Uri.parse(DRIVE_MOUNT_WIKI_LINK));
+      break;
+    case DriveMountUnsupportedAction.VIEW_ISSUE:
+      vs.env.openExternal(vs.Uri.parse(DRIVE_MOUNT_ISSUE_LINK));
+      break;
+  }
+}
+
+const DRIVE_MOUNT_PATTERN = /drive\.mount\(.*\)/;
```
Comment (kevineger) [inline src/jupyter/drive-mount-warning.ts]: Hopefully people don't write other code which looks like this. I think it'll be slim enough of a case not to be an issue. But unfortunately we (a) won't catch all with this (e.g. consider an aliased import) and (b) we might catch non-drive mount calls (rare...).

Should be fine!

---

We can be _slightly_ more precise with the regex tho.

```suggestion
const DRIVE_MOUNT_PATTERN = /drive\.mount\(.+\)/;
```

File: src/jupyter/drive-mount-warning.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    parsedJupyterMessage.header.msg_type === 'execute_request' &&
+    DRIVE_MOUNT_PATTERN.exec(parsedJupyterMessage.content.code)
+  ) {
+    await notifyDriveMountUnsupported(vs);
+  }
+}
+
+async function notifyDriveMountUnsupported(vs: typeof vscode): Promise<void> {
+  const selectedAction = await vs.window.showWarningMessage(
+    `drive.mount() is not supported by Colab VS Code extension at the moment, and we are actively working on supporting it. Please see [our wiki](${DRIVE_MOUNT_WIKI_LINK}) for workaround and [this issue](${DRIVE_MOUNT_ISSUE_LINK}) for progress.`,
```
Comment (kevineger) [inline src/jupyter/drive-mount-warning.ts]: Grammar + brevity

```suggestion
    `drive.mount is not currently supported in the extension. We're working on it! See the [wiki](${DRIVE_MOUNT_WIKI_LINK}) for workarounds and track this [issue](${DRIVE_MOUNT_ISSUE_LINK}) for progress.`,
```

File: src/jupyter/drive-mount-warning.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+const DRIVE_MOUNT_PATTERN = /drive\.mount\(.*\)/;
+const DRIVE_MOUNT_ISSUE_LINK =
+  'https://github.com/googlecolab/colab-vscode/issues/256';
+const DRIVE_MOUNT_WIKI_LINK =
+  'https://github.com/googlecolab/colab-vscode/wiki/Known-Issues-and-Workarounds#drivemount';
+
+enum DriveMountUnsupportedAction {
+  VIEW_ISSUE = 'View Issue',
+  VIEW_WORKAROUND = 'View Workaround',
```
Comment (kevineger) [inline src/jupyter/drive-mount-warning.ts]: Nit: _View_ may be redundant.

```suggestion
  VIEW_ISSUE = 'GitHub Issue',
  VIEW_WORKAROUND = 'Workaround',
```

File: src/jupyter/drive-mount-warning.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ * an execute request containing `drive.mount()`.
+ */
+export async function warnOnDriveMount(
+  vs: typeof vscode,
+  rawJupyterMessage: string,
+): Promise<void> {
+  const parsedJupyterMessage = JSON.parse(rawJupyterMessage) as {
+    header: { msg_type: string };
+    content: { code: string };
+  };
```
Comment (kevineger) [inline src/jupyter/drive-mount-warning.ts]: Isn't this `as` cast and raw access unsafe? Won't it cause unhandled errors? Shouldn't we use a type-guard, or better yet a Zod definition with safe parsing?

File: src/jupyter/drive-mount-warning.unit.test.ts [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+ * Copyright 2025 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import sinon from 'sinon';
+import { Uri } from 'vscode';
+import { newVsCodeStub, VsCodeStub } from '../test/helpers/vscode';
+import { warnOnDriveMount } from './drive-mount-warning';
+
+describe('warnOnDriveMount', () => {
```
Comment (kevineger) [inline src/jupyter/drive-mount-warning.unit.test.ts:13]: Related to cast / type-guard / parsing, add a test for empty data, or other data. We want to make sure we don't throw in that case. Should no-op.

File: src/jupyter/assignments.unit.test.ts [hunk 13]
```diff
@@ -1365,14 +1391,19 @@ describe('AssignmentManager', () => {
...
+  const { fetch: _f, WebSocket: _ws, ...c } = server.connectionInformation;
   return {
     ...server,
     connectionInformation: c,
   };
 }
 
-function stripFetches(servers: ColabAssignedServer[]): ColabAssignedServer[] {
-  return servers.map(stripFetch);
+function stripFetchesAndWebSockets(
```
Comment (kevineger) [inline src/jupyter/assignments.unit.test.ts]: Nit: could we name this `stripNetworkOverrides`?

File: src/jupyter/assignments.unit.test.ts [hunk 14]
```diff
@@ -496,6 +504,13 @@ describe('AssignmentManager', () => {
             }),
           });
         });
+
+        it('includes a custom WebSocket implementation', async () => {
+          const servers = await assignmentManager.getServers('extension');
+          assert.lengthOf(servers, 1);
+          const server = servers[0];
+          assert.isDefined(server.connectionInformation.WebSocket);
```
Comment (kevineger) [inline src/jupyter/assignments.unit.test.ts:504]: Similar to what's done above, can we guard that this attaches the headers we require in order to use this websocket?

File: src/jupyter/assignments.unit.test.ts [hunk 15]
```diff
@@ -863,6 +879,10 @@ describe('AssignmentManager', () => {
           }),
         });
       });
+
+      it('includes a custom WebSocket implementation', () => {
+        assert.isDefined(assignedServer.connectionInformation.WebSocket);
```
Comment (kevineger) [inline src/jupyter/assignments.unit.test.ts:875]: Same here.

**Review by kevineger — CHANGES_REQUESTED** [review]
Sorry it took me so long to review this, crazy week. Largely looking great!

File: src/jupyter/assignments.ts (hunk 5 above)
Comment (kevineger) [reply src/jupyter/assignments.ts]: Some other questions:

- Can `type SuperSend = WebSocketIsomorphic['send'];` be used to simplify code as well as make the code a bit more future proof?
- What about `...args` spreading?

File: package.json (hunk 1 above)
Comment (hjjackyang) [reply package.json]: Do we foresee supporting Web Extensions in the future? Perhaps it's better to be a bit future proof, and at the same time align with Jupyter extension implementation (which also uses `isomorphic-ws`)?

File: src/jupyter/assignments.ts (hunk 2 above)
Comment (hjjackyang) [reply src/jupyter/assignments.ts]: Done

File: src/jupyter/assignments.ts (hunk 3 above)
Comment (hjjackyang) [reply src/jupyter/assignments.ts]: Done

File: src/jupyter/drive-mount-warning.ts (hunk 11 above)
Comment (hjjackyang) [reply src/jupyter/drive-mount-warning.ts]: Done

File: src/jupyter/drive-mount-warning.unit.test.ts (hunk 12 above)
Comment (hjjackyang) [reply src/jupyter/drive-mount-warning.unit.test.ts:13]: You are right! New tests do fail and I added the proper type guard.

File: src/jupyter/assignments.unit.test.ts (hunk 13 above)
Comment (hjjackyang) [reply src/jupyter/assignments.unit.test.ts]: Done. Renamed to `stripNetworkOverride` for a single server and `stripNetworkOverrides` for multiple servers.

File: src/jupyter/assignments.ts (hunk 6 above)
Comment (hjjackyang) [reply src/jupyter/assignments.ts]: Done

File: src/jupyter/assignments.ts (hunk 5 above)
Comment (hjjackyang) [reply src/jupyter/assignments.ts]: Looks like I'm able to override just the implementation signature without the various overloads.

Re `type SuperSend = WebSocketIsomorphic['send']`, I'm not familiar with how to use it to simplify the code. Would be great if you can suggest an example. Thanks!

File: src/jupyter/drive-mount-warning.ts (hunk 7 above)
Comment (hjjackyang) [reply src/jupyter/drive-mount-warning.ts]: Done

File: src/jupyter/drive-mount-warning.ts (hunk 8 above)
Comment (hjjackyang) [reply src/jupyter/drive-mount-warning.ts]: Done

File: src/jupyter/drive-mount-warning.ts (hunk 9 above)
Comment (hjjackyang) [reply src/jupyter/drive-mount-warning.ts]: Done

File: src/jupyter/drive-mount-warning.ts (hunk 10 above)
Comment (hjjackyang) [reply src/jupyter/drive-mount-warning.ts]: Done

File: src/jupyter/assignments.ts (hunk 4 above)
Comment (hjjackyang) [reply src/jupyter/assignments.ts]: Done

File: src/jupyter/assignments.unit.test.ts (hunk 14 above)
Comment (hjjackyang) [reply src/jupyter/assignments.unit.test.ts:504]: I spent a couple of hours on this, but unable to make it work unfortunately. I wonder if it's actually worth testing this, considering our e2e tests will cover and ensure our custom WebSocket works as expected.

Comment (hjjackyang) [reply src/jupyter/assignments.unit.test.ts:504]: As discussed offline, I tried `sinon.stub(WebSocketIsomorphic.prototype, 'send')` and it didn't work either, so I went with the discussed solution of extracting out `colabProxyWebSocket` function and passing in `typeof WebSocketIsomorphic` for easier stubbing.

Please kindly review again!

File: src/jupyter/colab-proxy-web-socket.ts [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+import WebSocketIsomorphic from 'isomorphic-ws';
+import vscode from 'vscode';
+import {
+  COLAB_CLIENT_AGENT_HEADER,
+  COLAB_RUNTIME_PROXY_TOKEN_HEADER,
+} from '../colab/headers';
+import { warnOnDriveMount } from './drive-mount-warning';
+
+/**
+ * Returns a `WebSocket` class which extends `WebSocketIsomorphic`, adds Colab's
```
Comment (kevineger) [inline src/jupyter/colab-proxy-web-socket.ts]: Nit: consider `{@link ...}` comments where relevant.

File: src/jupyter/colab-proxy-web-socket.unit.test.ts [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+      address: string | URL | null,
+      protocols?:
+        | string
+        | string[]
+        | WebSocketIsomorphic.ClientOptions
+        | ClientRequestArgs,
+      options?: WebSocketIsomorphic.ClientOptions | ClientRequestArgs,
+    ) {
+      super(null); // Avoid real WS connection
+      if (address !== null) {
```
Comment (kevineger) [inline src/jupyter/colab-proxy-web-socket.unit.test.ts]: Remove so that `TestWebSocket` _always_ validates the headers are present?

File: src/jupyter/drive-mount-warning.ts [hunk 18]
```diff
@@ -10,45 +10,71 @@ import vscode from 'vscode';
...
   );
-  switch (selectedAction) {
-    case DriveMountUnsupportedAction.VIEW_WORKAROUND:
-      vs.env.openExternal(vs.Uri.parse(DRIVE_MOUNT_WIKI_LINK));
-      break;
-    case DriveMountUnsupportedAction.VIEW_ISSUE:
-      vs.env.openExternal(vs.Uri.parse(DRIVE_MOUNT_ISSUE_LINK));
-      break;
-  }
 }
```
Comment (kevineger) [inline src/jupyter/drive-mount-warning.ts:33]: Nit: did you consider Zod?

This may be more readable and robust:

```ts
export const ExecuteRequestSchema = z.object({
  header: z.object({
    msg_type: z.literal("execute_request"),
  }),
  content: z.object({
    code: z.string(),
  }),
});
```

```ts
export function isExecuteRequest(
  message: unknown,
): message is JupyterExecuteRequestMessage {
  return ExecuteRequestSchema.safeParse(message).success;
}
```

File: src/jupyter/drive-mount-warning.unit.test.ts [hunk 19]
```diff
@@ -23,20 +23,26 @@ describe('warnOnDriveMount', () => {
     });
 
     it('shows warning notification', async () => {
-      await warnOnDriveMount(vsCodeStub.asVsCode(), rawDriveMountMessage);
+      const showWarningMessageStub = vsCodeStub.window
+        .showWarningMessage as sinon.SinonStub;
+      showWarningMessageStub.resolves(undefined);
+
+      warnOnDriveMount(vsCodeStub.asVsCode(), rawDriveMountMessage);
+      await flush();
```
Comment (kevineger) [inline src/jupyter/drive-mount-warning.unit.test.ts]: In the cases where we expect "does nothing", the best we can do is `flush` then assert it wasn't called.

But where we're expecting the message, why not use proper async/await mechanisms? E.g. take a look at [this](https://github.com/googlecolab/colab-vscode/blob/84049083b68e7e3c78bd4c3ba6501455e99279ce/src/auth/flows/loopback.unit.test.ts#L148-L153https://github.com/googlecolab/colab-vscode/blob/84049083b68e7e3c78bd4c3ba6501455e99279ce/src/auth/flows/loopback.unit.test.ts#L148-L153) for reference.

File: src/jupyter/assignments.ts [hunk 20]
```diff
@@ -212,13 +213,17 @@ export class AssignmentManager implements vscode.Disposable {
...
-        },
-      }));
+      ).map((server) => {
+        const c = server.connectionInformation;
+        return {
+          ...server,
+          connectionInformation: {
+            ...c,
+            fetch: colabProxyFetch(c.token),
+            WebSocket: colabProxyWebSocket(this.vs, c.token),
```
Comment (kevineger) [inline src/jupyter/assignments.ts:223]: This is another situation where it'd be great if we had experiment config. I get a little worried / anxious going 0 to 100% for overriding the `WebSocket` impl. Not actionable, just mentioning.

File: package.json (hunk 1 above)
Comment (kevineger) [reply package.json]: It's very unlikely we'd ever ship as a non-Universal extension, `Web` nor platform specific.

File: src/jupyter/assignments.ts (hunk 5 above)
Comment (kevineger) [reply src/jupyter/assignments.ts]: You could simplify the member signature by referencing the super's sig like that.

File: package.json (hunk 1 above)
Comment (hjjackyang) [reply package.json]: Done

File: src/jupyter/assignments.ts (hunk 5 above)
Comment (hjjackyang) [reply src/jupyter/assignments.ts]: Please excuse my lack of TS knowledge. I'm still not sure what the syntax of override would look like with `type SuperSend = WebSocketIsomorphic['send']`. Could you provide an example snippet?

File: src/jupyter/colab-proxy-web-socket.ts (hunk 16 above)
Comment (hjjackyang) [reply src/jupyter/colab-proxy-web-socket.ts]: Done

File: src/jupyter/colab-proxy-web-socket.unit.test.ts (hunk 17 above)
Comment (hjjackyang) [reply src/jupyter/colab-proxy-web-socket.unit.test.ts]: Done

File: src/jupyter/drive-mount-warning.ts (hunk 18 above)
Comment (hjjackyang) [reply src/jupyter/drive-mount-warning.ts:33]: Done

File: src/jupyter/drive-mount-warning.unit.test.ts (hunk 19 above)
Comment (hjjackyang) [reply src/jupyter/drive-mount-warning.unit.test.ts]: What's the proper async/await mechanism here when `warnOnDriveMount` is a sync function and there is no promise to be awaited?

File: src/jupyter/assignments.ts (hunk 5 above)
Comment (kevineger) [reply src/jupyter/assignments.ts]: I was wondering if you could use the base member's signature typing to reduce definition clutter here. I don't know if you can or not.

But my other point is that I'm pretty sure you only need to `override` the implementing signature.

File: src/jupyter/drive-mount-warning.unit.test.ts (hunk 19 above)
Comment (kevineger) [reply src/jupyter/drive-mount-warning.unit.test.ts]: Something like the reference I linked. E.g.:

```ts
const warningShown = new Promise<void>((r) => {
  showWarningMessageStub.callsFake(() r.resolve());
  return Promise.resolve(true);
});

warnOnDriveMount(vsCodeStub.asVsCode(), rawDriveMountMessage);

await expect(warningShown).to.be.eventually.fulfilled;
```

File: src/jupyter/assignments.ts (hunk 5 above)
Comment (hjjackyang) [reply src/jupyter/assignments.ts]: SG! I've already updated it to only override the implementing signature.

File: src/jupyter/drive-mount-warning.unit.test.ts (hunk 19 above)
Comment (hjjackyang) [reply src/jupyter/drive-mount-warning.unit.test.ts]: Done
